File: web/dart/versions/version-1.dart
part of NetTango;

class Version1 {

  static void update(JsObject json) {
    Map<String, int> actionToId = new Map<String, int>();
    Map<int, int> blockIdToAttributeIdOffset = new Map<int, int>();
    int attributeId = 0;

    Function(JsObject) blockDefinitionHandler = (b) {
      if (!b.hasProperty("action")) { return; }

      int id = actionToId.length;
      b["id"] = id;
      String action = b["action"];
      actionToId[action] = id;

      blockIdToAttributeIdOffset[id] = attributeId;

      attributeId = addIdsToParamsAndProps(attributeId, b);
    };

    Function(JsObject) blockInstanceHandler = (b) {
      addIdToBlock(actionToId, blockIdToAttributeIdOffset, b);
    };

    VersionUtils.updateBlocks3(json, blockDefinitionHandler, blockInstanceHandler);

  }

  static int addIdsToParamsAndProps(int attributeId, JsObject b) {
    Function(JsArray) attributesHandler = (attributes) {
      attributeId = addIdsToAttributes(attributeId, attributes);
    };
    VersionUtils.updateBlockAttributes(b, attributesHandler);

    return attributeId;
  }

  static int addIdsToAttributes(int attributeId, JsArray attributes) {
    for (JsObject attribute in attributes) {
      attribute["id"] = attributeId;
      attributeId++;
    }
    return attributeId;
  }

  static void addIdToBlock(Map<String, int> actionToId, Map<int, int> blockIdToAttributeIdOffset, JsObject b) {
    if (!b.hasProperty("action")) { return; }

    String action = b["action"];
    if (actionToId.containsKey(action)) {
      int id = actionToId[action];
      b["id"] = id;
      addIdsToParamsAndProps(blockIdToAttributeIdOffset[id], b);
    }
  }

}

File: src/versions/version-1.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import { VersionUtils } from "./version-utils"

class Version1 {

  static update(json: any): any {
    const actionToId: Map<string, number> = new Map()
    const blockIdToAttributeIdOffset: Map<number, number> = new Map()
    var attributeId = 0

    const blockDefinitionHandler = (b: any) => {
      if (!b.hasOwnProperty("action")) { return }

      const id = actionToId.size
      b["id"] = id
      const action: string = b["action"]
      actionToId.set(action, id)

      blockIdToAttributeIdOffset.set(id, attributeId)

      attributeId = Version1.addIdsToParamsAndProps(attributeId, b)
    }

    const blockInstanceHandler = (b: any) => {
      Version1.addIdToBlock(actionToId, blockIdToAttributeIdOffset, b)
    }

    VersionUtils.updateBlocks3(json, blockDefinitionHandler, blockInstanceHandler)

  }

  static addIdsToParamsAndProps(attributeId: number, b: any): number {
    const attributesHandler = (attributes: any[]) => {
      attributeId = Version1.addIdsToAttributes(attributeId, attributes)
    }
    VersionUtils.updateBlockAttributes(b, attributesHandler)

    return attributeId
  }

  static addIdsToAttributes(attributeId: number, attributes: any[]): number {
    for (var attribute of attributes) {
      attribute["id"] = attributeId
      attributeId++
    }
    return attributeId
  }

  static addIdToBlock(actionToId: Map<string, number>, blockIdToAttributeIdOffset: Map<number, number>, b: any): void {
    if (!b.hasOwnProperty("action")) { return }

    const action: string = b["action"]
    if (actionToId.has(action)) {
      const id = actionToId.get(action)!
      b["id"] = id
      Version1.addIdsToParamsAndProps(blockIdToAttributeIdOffset.get(id)!, b)
    }
  }

}

export { Version1 }